client/telemetry/src/lib.rs

```diff
@@ -16,339 +16,406 @@
-	// `Multiaddr`.
-	Err(parse_error)
-}
-
-/// Log levels.
-pub const SUBSTRATE_DEBUG: &str = "9";
-pub const SUBSTRATE_INFO: &str = "0";
-
-pub const CONSENSUS_TRACE: &str = "9";
-pub const CONSENSUS_DEBUG: &str = "5";
-pub const CONSENSUS_WARN: &str = "4";
-pub const CONSENSUS_INFO: &str = "1";
-
-/// Telemetry object. Implements `Future` and must be polled regularly.
-/// Contains an `Arc` and can be cloned and pass around. Only one clone needs to be polled
-/// regularly and should be polled regularly.
-/// Dropping all the clones unregisters the telemetry.
-#[derive(Clone)]
-pub struct Telemetry {
-	inner: Arc<Mutex<TelemetryInner>>,
-	/// Slog guard so that we don't get deregistered.
-	_guard: Arc<slog_scope::GlobalLoggerGuard>,
-}
-
-/// Behind the `Mutex` in `Telemetry`.
-///
-/// Note that ideally we wouldn't have to make the `Telemetry` cloneable, as that would remove the
-/// need for a `Mutex`. However there is currently a weird hack in place in `sc-service`
-/// where we extract the telemetry registration so that it continues running during the shutdown
-/// process.
-struct TelemetryInner {
-	/// Worker for the telemetry. `None` if it failed to initialize.
-	worker: Option<worker::TelemetryWorker>,
-	/// Receives log entries for them to be dispatched to the worker.
-	receiver: mpsc::Receiver<async_record::AsyncRecord>,
-}
+	/// Run the telemetry worker.
+	///
+	/// This should be run in a background task.
+	pub async fn run(self) {
+		let Self {
+			receiver,
+			sender: _sender,
+			mut init_receiver,
+			init_sender: _init_sender,
+			transport,
+		} = self;
+
+		let mut node_map: HashMap<Id, Vec<(u8, Multiaddr)>> = HashMap::new();
+		let mut node_args: HashMap<
+			Multiaddr,
+			(Vec<ConnectionMessage>, Vec<ConnectionNotifierSender>),
+		> = HashMap::new();
+		let mut existing_nodes: HashSet<Multiaddr> = HashSet::new();
+
+		// initialize the telemetry nodes
+		init_receiver.close();
+		while let Some((id, endpoints, mut connection_message, mut connection_notifiers_rx)) =
+			init_receiver.next().await
+		{
+			let endpoints = endpoints.0;
+
+			connection_notifiers_rx.close();
+			let connection_notifier_senders: Vec<_> = connection_notifiers_rx.collect().await;
+
+			connection_message.insert("msg".into(), "system.connected".into());
+
+			for (addr, verbosity) in endpoints {
+				node_map
+					.entry(id.clone())
+					.or_insert_with(Vec::new)
+					.push((verbosity, addr.clone()));
+				existing_nodes.insert(addr.clone());
+				let mut obj = serde_json::Map::new();
+				obj.insert("id".to_string(), id.into_u64().into());
+				obj.insert("payload".to_string(), connection_message.clone().into());
+				let (connection_messages, connection_notifiers) = node_args
+					.entry(addr.clone())
+					.or_insert_with(|| (Vec::new(), Vec::new()));
+				connection_messages.push(obj);
+				connection_notifiers.extend(connection_notifier_senders.clone());
```

Comment:
```suggestion
				connection_notifiers.extend(connection_notifier_senders);
```

Reply:
This one is a mistake. The connection_notifier_senders is in a loop